### ISSUE TITLE
Enhancement to support FTP connection with NAT scenario

### ISSUE THREAD

**Issue by yugandhar2587** [python/cpython#127478]
# Feature or enhancement

### Proposal:

Some of new FTP servers are responding with 229 Entering Extended passive mode response when PASV command is sent.. Because of this sendcmd('PASV') is sending error though it's a accepted response. Tested with filezilla manually where it's accepting the same. So we need to make changes in ftplib to accept & parse Extended mode with 'PASV' command. Please help to review and merge.

### Has this already been discussed elsewhere?

This is a minor feature, which does not need previous discussion elsewhere

### Linked PRs
* gh-127477

**Comment by picnixz** [comment]
Some questions:

- Does the `ftplib` module exactly follow the corresponding RFC? 
- Are those new servers following the RFC or is this some custom behaviour?

Depending on the answer, we might or not accept such feature (though I can see why you would like to have it). It's just that if we say that we are strictly implementing a protocol and that this protocol is not followed by some FTP servers, then maybe the issue does not necessarily lie with Python (again I'm just confirming whether we can or not accept this feature request; personally I'd say yes).

**Comment by yugandhar2587** [comment]
> Some questions:
> 
> * Does the `ftplib` module exactly follow the corresponding RFC?
> * Are those new servers following the RFC or is this some custom behaviour?
> 
> Depending on the answer, we might or not accept such feature (though I can see why you would like to have it). It's just that if we say that we are strictly implementing a protocol and that this protocol is not followed by some FTP servers, then maybe the issue does not necessarily lie with Python (again I'm just confirming whether we can or not accept this feature request; personally I'd say yes if it's ok with the RFC and if it's a relevant response code).

Hi @picnixz ,

I observed this behavior from FTP server particularly in NAT'ing scenario. Seems ftplib is always making passive connection using PASV command before starting data transfer. As stated in [RFC-2428](https://datatracker.ietf.org/doc/html/rfc2428), EPSV command should be used for IPv6 and NAT environments. So I feel we could consider this as an enhancement. Please help to review and revert.

**Comment by picnixz** [comment]
I wanted a specific section in the RFC but I found it (https://datatracker.ietf.org/doc/html/rfc2428#section-3):

> The response code for entering passive mode using an extended address MUST be 229.

To my understanding, this is only needed for the EPSV command, not the PASV command. AFAIK, the relevant specs are given in https://www.rfc-editor.org/rfc/rfc959 but I can't find anything except the diagram on Figure 3 which indicates a 227 response code and not a 229. So, we should only accept such code if we are sending an EPSV command, not if we are using a PASV command only. 

So I really don't think we should accept 229 as a response code for PASV commands.

**Comment by yugandhar2587** [comment]
> I wanted a specific section in the RFC but I found it (https://datatracker.ietf.org/doc/html/rfc2428#section-3):
> 
> > The response code for entering passive mode using an extended address MUST be 229.
> 
> To my understanding, this is only needed for the EPSV command, not the PASV command. AFAIK, the relevant specs are given in https://www.rfc-editor.org/rfc/rfc959 but I can't find anything except the diagram on Figure 3 which indicates a 227 response code and not a 229. So, we should only accept such code if we are sending an EPSV command, not if we are using a PASV command only.
> 
> So I really don't think we should accept 229 as a response code for PASV commands.
> 
> EDIT: The protocol really expects 227 to be used for PASV and 229 to be used for EPSV, we can't make them interchangeable IMO.

https://datatracker.ietf.org/doc/html/rfc2428#section-4 is stating to use EPSV for IPv6 or NAT scenario. But ftplib is always sending PASV command as part of ntranfercmd.
```

    def ntransfercmd(self, cmd, rest=None):
        """Initiate a transfer over the data connection.

        If the transfer is active, send a port command and the
        transfer command, and accept the connection.  If the server is
        passive, send a pasv command, connect to it, and start the
        transfer command.  Either way, return the socket for the
        connection and the expected size of the transfer.  The
        expected size may be None if it could not be determined.

```
Moreover figuring out whether given host name is of/supports IPv6 or it's private IP (in case of NAT) is bit tricky. So we can accept 229 Extended Passive mode as PASV command output. Further more makepasv() should return  host, port, which will hold true with the current enhancement when server is responding with 229 with PASV command.

**Comment by yugandhar2587** [comment]
> AFAICT, `makepasv()` checks whether we are using IPv4 or IPv6 via `self.af == socket.AF_INET`. The IPv6 scenario is already handled in the `else` branch (so we already support EPSV as it is also used in IPv6). I'm not against supporting 229, but what I'm against is that we expect PASV to return 229 (whether it's NAT or not, this shouldn't happen: when you use PASV, you should return 227; when you use EPSV, you should return 229).
> 
> Now, is there a way that we're in the NAT case *and3 that `self.af == socket.AF_INET`?

yes you are right. We are currently hitting the NAT case with `self.af == socket.AF_INET`?

**Comment by picnixz** [comment]
Oh, then we have some solutions:

- We try to correctly detect the NAT case (i.e., we try to follow the RFC as much as possible). This may be hard and I don't know how feasible it is.
- According to the RFC, the NAT case should be using 'EPSV' right? So, instead of accepting 229 as a valid 'PASV' response code, I suggest resending 'EPSV' if 'PASV' failed (having '229' returned for PASV seems wrong to me, even in the NAT case).

**Comment by picnixz** [comment]
> do you mean for any error we can retry EPSV command to make passive connection or only for 229 Response ?

If an error with `PASV` happens -> try `EPSV` + check for 229. I don't think it makes sense to expect 229 from `PASV` (this is an issue on the server side in this case since it doesn't follow the RFC).

**Comment by gpshead** [comment]
The addage to keep in mind in network programming is: "Be strict on what you produce, and lenient in what you accept."

If you are indeed seeing a server reply with a `229` normally reserved only for `EPSV` response to a `PASV` command instead of the expected `227` it seems natural to assume that we should just accept the 229 and go with it as the server at that point should be presumed to have bound to a port and is awaiting our data connection.  Such a server is not strictly adhering to the ftp RFC 959 and RFC 2428 but its response is still comprehensible.

So I would change your current iteration of the PR from

```python
            try:
                untrusted_host, port = parse227(self.sendcmd('PASV'))
            except error_reply:
                untrusted_host, port = parse229(self.sendcmd('EPSV'), self.sock.getpeername())
```

into something more like:

```python
            pasv_response = self.sendcmd('PASV')
            try:
                untrusted_host, port = parse227(pasv_response)
            except error_reply:
                untrusted_host, port = parse229(pasv_response, self.sock.getpeername())
```

**Comment by yugandhar2587** [comment]
> The addage to keep in mind in network programming is: "Be strict on what you produce, and lenient in what you accept."
> 
> If you are indeed seeing a server reply with a `229` normally reserved only for `EPSV` response to a `PASV` command instead of the expected `227` it seems natural to assume that we should just accept the 229 and go with it as the server at that point should be presumed to have bound to a port and is awaiting our data connection. Such a server is not strictly adhering to the ftp RFC 959 and RFC 2428 but its response is still comprehensible.
> 
> So I would change your current iteration of the PR from
> 
> ```python
>             try:
>                 untrusted_host, port = parse227(self.sendcmd('PASV'))
>             except error_reply:
>                 untrusted_host, port = parse229(self.sendcmd('EPSV'), self.sock.getpeername())
> ```
> 
> into something more like:
> 
> ```python
>             pasv_response = self.sendcmd('PASV')
>             try:
>                 untrusted_host, port = parse227(pasv_response)
>             except error_reply:
>                 untrusted_host, port = parse229(pasv_response, self.sock.getpeername())
> ```

Thanks for the confirmation. Changed as suggested.

**Comment by simonepelosi** [comment]
I have submitted a PR that addresses this: https://github.com/python/cpython/pull/150236

The change makes `makepasv()` try EPSV before PASV on IPv4 connections, falling back to PASV automatically if the server doesn't support it.
